File: hard/940-distinct-subsequences-ii/solution.java
/**
 * 940. Distinct Subsequences II
 * Time: O(n)
 * Space: O(26)
 */
class Solution {
    private static final int MOD = 1_000_000_007;

    public int distinctSubseqII(String s) {
        // ending[c] = distinct subsequences ending in character c
        long[] ending = new long[26];
        long total = 0;

        for (int i = 0; i < s.length(); i++) {
            int c = s.charAt(i) - 'a';
            // append c to every existing subsequence, plus "c" alone
            long updated = (total + 1) % MOD;
            // overwrite the bucket: earlier duplicates are regenerated here
            total = (total - ending[c] + updated + MOD) % MOD;
            ending[c] = updated;
        }

        return (int) (total % MOD);
    }
}

File: hard/940-distinct-subsequences-ii/solution.py
class Solution:
    def distinctSubseqII(self, s: str) -> int:
        """
        940. Distinct Subsequences II
        Time: O(n)
        Space: O(26)
        """
        MOD = 10**9 + 7

        # ending[c] = distinct subsequences ending in character c
        ending = [0] * 26
        total = 0

        for ch in s:
            c = ord(ch) - 97
            # append c to every existing subsequence, plus "c" alone
            updated = (total + 1) % MOD
            # overwrite the bucket: earlier duplicates are regenerated here
            total = (total - ending[c] + updated) % MOD
            ending[c] = updated

        return total % MOD
